Highlight by-name parameters at their creation

mlangc opened the pull request:
This should fix #1002340 by treating call-by-name parameters at their creation similar to implicit conversions.

mlangc commented:
Hmm, I don't think that this is related to my changes:
```
[ERROR] Failed to execute goal on project org.scala-ide.sbt.full.library: Could not resolve dependencies for project org.scala-ide:org.scala-ide.sbt.full.library:bundle:4.1.0-SNAPSHOT: The following artifacts could not be resolved: org.scala-lang:scala-library:jar:2.11.5-SNAPSHOT, org.scala-lang:scala-compiler:jar:2.11.5-SNAPSHOT, org.scala-lang:scala-reflect:jar:2.11.5-SNAPSHOT: Could not find artifact org.scala-lang:scala-library:jar:2.11.5-20150108.035808-68 in moxie-everything (https://moxie.typesafe.com:8497/nexus/content/groups/everything) -> [Help 1]
```

mlangc commented:
Hmm, I guess then I wait till this is fixed in master.

sschaef commented on org.scala-ide.sdt.core.tests/src/org/scalaide/core/semantichighlighting/classifier/RegionParser.scala:
```diff
@@ -2,10 +2,44 @@ package org.scalaide.core.semantichighlighting.classifier
+  }
+
+  object EmbeddedSubstr {
+    implicit def wrapAsEmbeddedSubstring(str: String) = EmbeddedSubstr(str)
```
implicit conversions produce a warning, please enable them with `import language.implicitConversions`

sschaef replied:
This implicit can be in scope of `RegionParser`.

mlangc replied:
Hmm, but my intention was that the conversion is available whenever somebody imports `EmbeddedSubstr`. Do you think that this is a bad idea?

sschaef replied:
When someone imports `RegionParser` they already want to use `EmbeddedSubstr`. Forcing them to do another import is not the right thing to do.

mlangc replied:
This already works because the implicit conversion is defined in the companion object; try it yourself:

```
scala> import org.scalaide.core.semantichighlighting.classifier.RegionParser._
scala> val x: EmbeddedSubstr = "3"
res0: EmbeddedSubstr(3,,)
```

sschaef commented on org.scala-ide.sdt.core.tests/src/org/scalaide/core/semantichighlighting/classifier/RegionParser.scala:
```diff
@@ -2,10 +2,44 @@ package org.scalaide.core.semantichighlighting.classifier
+  /**
+   * Extracts the regions marked by the given substrings.
+   */
+  def substrRegions(test: String, substrs: EmbeddedSubstr*): Map[IRegion, EmbeddedSubstr] = {
```
Please describe more clearly how this method is supposed to work. See `delimitedRegions` for an example.

mlangc replied:
See 687481aa13c0c40c73dbfa5bfe61a9af2c9cf3f4

sschaef commented on org.scala-ide.sdt.core.tests/src/org/scalaide/core/semantichighlighting/classifier/RegionParser.scala:
```diff
@@ -2,10 +2,44 @@ package org.scalaide.core.semantichighlighting.classifier
   /**
+   * This class represents a substring with an optional prefix and suffix.
+   */
+  case class EmbeddedSubstr(str: String, prefix: String = "", suffix: String = "") {
```
Try to mark all classes that are not intended to be subclassed with `final`.

sschaef commented on org.scala-ide.sdt.core.tests/src/org/scalaide/core/semantichighlighting/classifier/RegionParser.scala:
```diff
@@ -2,10 +2,44 @@ package org.scalaide.core.semantichighlighting.classifier
+   * This class represents a substring with an optional prefix and suffix.
+   */
+  case class EmbeddedSubstr(str: String, prefix: String = "", suffix: String = "") {
+    private[RegionParser] val searchString = prefix + str + suffix
```
The scope specifier is unnecessary, the fields are public anyway.

mlangc replied:
Removing the specifier leads to a compile error.

sschaef replied:
That is weird. Which error message do you get?

mlangc replied:
RegionParser.scala:51: error: value searchString in class EmbeddedSubstr cannot be accessed in org.scalaide.core.semantichighlighting.classifier.RegionParser.EmbeddedSubstr

sschaef commented on org.scala-ide.sdt.core.tests/src/org/scalaide/ui/internal/editor/decorators/params/byname/CallByNameParamAtCreationPresenterTest.scala:
```diff
@@ -0,0 +1,175 @@
+import org.scalaide.CompilerSupportTests
+import org.scalaide.core.semantichighlighting.classifier.RegionParser
+import org.scalaide.core.semantichighlighting.classifier.RegionParser.EmbeddedSubstr
+import org.scalaide.core.semantichighlighting.classifier.RegionParser.EmbeddedSubstr.wrapAsEmbeddedSubstring
```
duplicated imports, probably inserted by organize imports.

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/params/byname/CallByNameParamAtCreationAnnotation.scala:
```diff
@@ -0,0 +1,13 @@
+  val ID = "scala.tools.eclipse.semantichighlighting.callByNameParam.creationAnnotation"
+}
+
+class CallByNameParamAtCreationAnnotation(text: String)
```
missing `final`

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/params/byname/CallByNameParamAtCreationAnnotation.scala:
```diff
@@ -0,0 +1,13 @@
+package org.scalaide.ui.internal.editor.decorators.params.byname
```
a single package `bynameparams` seems to be better to me.

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/params/byname/CallByNameParamAtCreationHighlightingParticipant.scala:
```diff
@@ -0,0 +1,8 @@
+
+import org.scalaide.core.extensions.SemanticHighlightingParticipant
+
+class CallByNameParamAtCreationHighlightingParticipant extends
```
I think with this one you win the price for introducing the longest fully qualified name. ;)

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/params/byname/CallByNameParamAtCreationPresenter.scala:
```diff
@@ -0,0 +1,96 @@
+import org.scalaide.ui.internal.preferences.CallByNameParamCreationPreferencePage
+import org.scalaide.core.IScalaPlugin
+
+class CallByNameParamAtCreationPresenter(sourceViewer: ISourceViewer) extends
```
missing `final`

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/params/byname/CallByNameParamAtCreationPresenter.scala:
```diff
@@ -0,0 +1,96 @@
+}
+
+object CallByNameParamAtCreationPresenter extends HasLogger {
+  case class Cfg(firstLineOnly: Boolean)
```
A case class with one parameter and no super trait seems unnecessary, returning the boolean directly should be sufficient.

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/params/byname/CallByNameParamAtCreationPresenter.scala:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        def processArgs(funTpe: compiler.Type, args: List[compiler.Tree]): Map[Annotation, Position] = {
+          if (funTpe.params.size != args.size /* <- possible for faulty code */) {
```
I don't understand this comment. Can you give a concrete example?

mlangc replied:
I've just expanded the comment; I hope it is clearer now.

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/params/byname/CallByNameParamAtCreationPresenter.scala:
```diff
@@ -0,0 +1,96 @@
+    Cfg(prefStore.getBoolean(CallByNameParamCreationPreferencePage.P_FIRST_LINE_ONLY))
+  }
+
+  def annotations(compiler: IScalaPresentationCompiler, scu: ScalaCompilationUnit, sourceFile: SourceFile, cfg: Cfg): Map[Annotation, Position] = {
```
`annotations` is not a good method name. And why is this method part of the object?

sschaef replied:
Ok, I see why it is part of the object. Public definitions need Scaladoc then. Explain what it is doing, what does the return type mean etc.

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/params/byname/CallByNameParamAtCreationPresenter.scala:
```diff
@@ -0,0 +1,96 @@
+          new CallByNameParamAtCreationAnnotation(s"Call-by-name parameter creation: () => $txt")
+        }
+
+        def toPositioin(arg: compiler.Tree, txt: String): Position = {
```
spelling: Positioin

sschaef replied:
This method seems to be duplicated from the implicit highlighter. Decide by yourself if it is worth to extract it to a helper.

mlangc replied:
You got me on this; I would have already done so if I had a clear idea where such a helper should reside. Do you have any suggestions?

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/preferences/BasicFieldEditorPreferencePage.scala:
```diff
@@ -0,0 +1,20 @@
+import org.scalaide.core.IScalaPlugin
+import org.eclipse.jface.preference.BooleanFieldEditor
+
+abstract class BasicFieldEditorPreferencePage(description: String) extends FieldEditorPreferencePage with IWorkbenchPreferencePage {
```
missing Scaladoc

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/preferences/CallByNameParamCreationPreferencePage.scala:
```diff
@@ -0,0 +1,44 @@
+}
+
+object CallByNameParamCreationPreferencePage {
+  val BASE = "scala.tools.eclipse.ui.preferences.callByNameParamCreation"
```
Please inline `BASE`, this makes it possible to grep for the values.

sschaef replied:
Also prefer the `UpperCamelCase` notation.

mlangc replied:
I prefer `CamelCase`, but it seems to me that most constants in Scala-IDE are `CAPITALIZED_WITH_UNDERSCORE` (see org.scalaide.ui.syntax.ScalaSyntaxClasses for a bunch of examples). I don't think that is a good idea to mix conventions.

As for inlining `BASE`: Here I'm also just doing things like they are done by other pages (like `org.scalaide.ui.internal.preferences.OrganizeImportsPreferencesPage`, `org.scalaide.ui.internal.preferences.EditorPreferencePage`, ...). I think in the end it is a matter of taste what you prefer, as both approaches have advantages and disadvantages that cannot be easily compared in an objective manner.

sschaef replied:
In the past scala-ide contributors agreed to use CamelCase (if I remember correctly) and to inline constant names due to better grep'ability. Don't rely too much on the design of older code - a lot of it is broken and needs to be refactored.

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/preferences/PrefPageUtils.scala:
```diff
@@ -0,0 +1,23 @@
+import org.eclipse.ui.dialogs.PreferencesUtil
+import org.eclipse.swt.events.SelectionEvent
+
+private object PrefPageUtils {
```
Please use `SWTUtils`.

sschaef commented:
I'm finished with my review. Thanks for this PR, it is very useful!

Some points:
- Scaladoc at public definitions is a must have when they are not obvious getters.
- A more detailed commit message would be nice. For example explain why you chose to implement this feature through an editor extension instead of extending semantic highlighting. Explain which features are added/removed/changed (pref page added, implicit pref page refactored, etc.). It is also useful to move separate changes (the implicit pref page refactoring for example) to standalone commits, makes reviewing easier.

mlangc commented:
Thanks a lot for the review! I'm going to address these points at the weekend.

dragos commented:
retest this please

dragos commented:
(failures were transient, 2.11.5-SNAPSHOT is still there)

dragos commented:
Tests failed now because of a merge conflict.

mlangc commented:
Hmm, the build still fails grabbing org.scala-lang:scala-library:jar:2.11.5-SNAPSHOT... is there something that has to be done on my part?

mlangc commented:
I guess I really should look into the "flakyness" of `NamePrinterTest` sometime soon...

mlangc commented:
@sschaef  Well, I think I've addressed all issues except for extracting `CallByNameParamAtCreationPresenter.findByNameParamCreations...toPosition(arg: compiler.Tree, txt: String)` (see the question above). Please tell me if I missed anything else.

dragos commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/bynameparams/CallByNameParamAtCreationPresenter.scala:
```diff
@@ -0,0 +1,103 @@
+          sourceFile.content.view(arg.pos.start, arg.pos.end).mkString("")
+        }
+
+        def toAnnotation(arg: compiler.Tree, txt: String): Annotation = {
```
`arg` is not used.

mlangc replied:
Thanks for spotting this; I've already fixed this locally.

dragos commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/bynameparams/CallByNameParamAtCreationPresenter.scala:
```diff
@@ -0,0 +1,103 @@
+      traverser.result
+    }
+
+    compiler.askLoadedTyped(sourceFile, false).get match {
```
This shouldn't be necessary. It's already done for you in the superclass [BaseSemanticAction.apply](https://github.com/scala-ide/scala-ide/blob/master/org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/BaseSemanticAction.scala#L120).

mlangc replied:
You mean that checking the return value is not necessary? Or should I use another method to obtain the tree?

dragos replied:
Hm, sorry, I misread the definition of `findAll`. I think it's a mistake that it doesn't pass the tree, since it obtains it in `BaseSemanticAction`, hence forcing all implementers to re-post the request. Hopefully the tree didn't change in the meantime, so it hits the presentation compiler cache.

sschaef commented on org.scala-ide.sdt.core.tests/src/org/scalaide/core/semantichighlighting/classifier/RegionParser.scala:
```diff
@@ -2,10 +2,68 @@ package org.scalaide.core.semantichighlighting.classifier
+   * res0: Map(offset: 0, length: 1 -> EmbeddedSubstr(x,, ), offset: 6, length: 1 -> EmbeddedSubstr(x,(,)))
+   * }}}
+   */
+  def substrRegions(test: String, substrs: EmbeddedSubstr*): Map[IRegion, EmbeddedSubstr] = {
```
I think a solution similar to `delimitedRegions` would result in a better test suite. You can just surround a block by a matching pair of signs:
```scala
f([5])
def f(param: => Int) = param
```
Even nesting works:
```scala
f([f([5])])
def f(param: => Int) = param
```
You are just not allowed to use the pair signs anymore, but that shouldn't be a problem.

mlangc replied:
In my opinion there are several advantages of `substrRegions` compared to `delimitedRegions`:

1. There is not need to obscure the test cases with markers.
2. You don't need templates (see AbstractSymbolClassifierTest.checkSymbolClassification) that result in duplication.
3. It's simpler.

I didn't think this through, because I don't think it's worth the effort anyway, but my suspicion is that `AbstractSymbolClassifierTest.checkSymbolClassification` and all tests that use it could get rid of the `locationTemplate` parameter and therefore a significant amount of duplication if they where refactored to use `substrRegions`.

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/preferences/ImplicitsPreferencePage.scala:
```diff
@@ -13,46 +13,33 @@ import org.eclipse.jdt.internal.ui.preferences.PreferencesMessages
-  val P_ITALIC = BASE + "text.italic"
-  val P_CONVERSIONS_ONLY = BASE + "conversions.only"
-  val P_FIRST_LINE_ONLY  = BASE + "firstline.only"
+  val PActive = "scala.tools.eclipse.ui.preferences.implicit.enabled"
```
I don't link the `P` prefixes. When you are already in refactoring you could also remove it but its ok nonetheless.

mlangc replied:
I'll happily rename not only these, but lot's of additional constants as soon as we got a working renaming feature in the GUI.

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/decorators/bynameparams/CallByNameParamAtCreationPresenter.scala:
```diff
@@ -0,0 +1,103 @@
+}
+
+object CallByNameParamAtCreationPresenter extends HasLogger {
+  final case class Cfg(firstLineOnly: Boolean)
```
`Cfg` is still used

mlangc replied:
I'll take care of this in the next few days (maybe I'll manage tomorrow).

sschaef commented:
It just came to my mind that your test cases are probably unstable because they lack a package declaration. Please add a unique one to each of them, then we are on the save side.

mlangc commented:
I'm away during the weekend, but I'll take care of the package declarations the next week.

mlangc commented:
Ahh, sorry, I guess should not commit when in a hurry; I've kept my implementation as it is more robust, but I'm not religious about that.

sschaef commented:
LGTM

sschaef commented:
It is a great feature, I like it so much! But I already found a bug: https://www.assembla.com/spaces/scala-ide/tickets/1002381-highlight-call-by-name-parameter-does-not-work-for-partially-applied-function#/activity/ticket: ;)
